### ISSUE TITLE
Dart.next: <Primitive>.fromEnvironment is a total misnomer

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#27585]
In my mind, the following would be equivalent:

```dart
const String.fromEnvironment('FOO') == Platform.environment['FOO']
```

It's not though, because the former is a `dart -d` argument, and the latter is actually an exported environment variable. I realize this is a big breaking API, but maybe we have some planned anyway.

My recommendation is `fromDefine`.

**Comment by lrhn** [comment]
This is unlikely to change at the current time. We'll consider if it's worth changing for 2.0.

**Comment by fsc8000** [comment]
There is always the possibility to add a new short form without breaking any existing uses of the API.

**Comment by matanlurey** [comment]
If you'll accept a PR, I'd be happy to add it as an alias for now, i.e.

```dart
const factory String.fromDefine = String.fromEnvironment;
```

And we could mark the other one as deprecated.

**Comment by bwilkerson** [comment]
If you do this, please let the analyzer team know. We have special handling of the `fromEnvironment` constructors to support constant expression evaluation and might need to add special handling of the alias (it ought to work without additional support, but we'd need to verify that).

**Comment by matanlurey** [comment]
@lrhn Are you OK with the above? If so, I'll send a PR.

**Comment by lrhn** [comment]
I appreciate the idea, but I don't think it's worth it to do this change now, the benefit is there, but it's not that big. There are lots of small things that we *can* change, but we should also not change too much - as usual, it's a trade-off.

**Comment by matanlurey** [comment]
@lrhn @floitschG Any thought about this for 2.0?

Even just creating an alias and deprecating `fromEnvironment` would help a lot.

**Comment by blagoev** [comment]
Indeed. This API (`string.fromEnvironment`) is quite convenient. And also named like that is totally unclear it can not read environment variables.

**Comment by FelixZY** [comment]
Bumping this. The distinction between "build time variable" and "run time variable" is important. Searching for "dart environment variables" almost exclusively yields results for `String.fromEnvironment` rather than `Platform.environment`.

The separation should be made much clearer, both syntactically and in documentation.

[Documentation](https://api.dart.dev/stable/2.18.5/dart-core/String/String.fromEnvironment.html) on `String.fromEnvironment` for Dart 2.18.5 state that `fromEnvironment` uses "environment declarations" and that:

> Environment declarations are provided by the surrounding system compiling or running the Dart program.

The term "environment declaration" nearly sounds interchangeable with "environment variable" to the unaware and the formulation "[...] or running the Dart program" gives the appearance of equivalent functionality.
